components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,167 @@
+        debug("Got a list of Supervisor terminals: $terminals")
+        connectTasksToTerminals(tasks, terminals)
+    }
+
+    private fun connectTasksToTerminals(
+            tasks: List<TaskStatus>,
+            terminals: MutableList<TerminalOuterClass.Terminal>
+    ) = runInEdt {
+            if (tasks.isEmpty()) {
+                backendTerminalManager.createNewSharedTerminal("Gitpod", "Terminal")
+            } else {
+                val aliasToTerminalMap: MutableMap<String, TerminalOuterClass.Terminal> = mutableMapOf()
+
+                for (terminal in terminals) {
+                    val terminalAlias = terminal.alias
+                    aliasToTerminalMap[terminalAlias] = terminal
+                }
+
+                for (task in tasks) {
+                    val terminalAlias = task.terminal
+                    val terminal = aliasToTerminalMap[terminalAlias]
+
+                    if (terminal != null) {
+                        createSharedTerminal(terminal)
+                    }
+                }
+            }
+        }
+
+    private fun getTerminalToolWindowRegisteredEvent(): CompletableFuture<Void> {
+        val completableFuture = CompletableFuture<Void>()
+
+        val messageBusConnection = project.messageBus.connect()
+
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener = object : ToolWindowManagerListener {
+            override fun toolWindowsRegistered(ids: MutableList<String>, toolWindowManager: ToolWindowManager) {
+                if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                    debug("TerminalToolWindow got registered!")
+                    completableFuture.complete(null)
+                    messageBusConnection.disconnect()
+                }
+            }
+        }
+
+        messageBusConnection.subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
+
+        return completableFuture
+    }
+
+    private fun getSupervisorTasksList(): CompletableFuture<List<TaskStatus>> {
+        val completableFuture = CompletableFuture<List<TaskStatus>>()
+        val taskStatusRequest = TasksStatusRequest.newBuilder().setObserve(true).build()
+        val taskStatusResponseObserver = object : ClientResponseObserver<TasksStatusRequest, TasksStatusResponse> {
+            override fun beforeStart(request: ClientCallStreamObserver<TasksStatusRequest>) {
+                lifetime.onTerminationOrNow {
+                    request.cancel(null, null)
+                }
+            }
+
+            override fun onNext(response: TasksStatusResponse) {
+                debug("Received task list: ${response.tasksList}")
+
+                for (task in response.tasksList) {
+                    if (task.state === TaskState.opening) {
+                        return
+                    }
+                }
+
+                this.onCompleted()
```

Comment:
I still don't understand how it works. `onCompleted` method is overriden and does nothing but print a message. Should not we call here `request.cancel` as well to stop streaming which wil lead to `onCompleted` eventually? Or maybe `onCompleted` should call `super.onCompleted` to trigger some default behaviour which stops streaming?

Could you point me please to relevant docs? 🙏

Reply:
Sure, I found the documentation [here](https://grpc.io/docs/languages/java/basics/): "We use the response observer’s `onCompleted()` method to specify that we’ve finished dealing with the RPC."

And I agree that it needs to be changed to `super.onCompleted` for it to make effect. Although I couldn't find any usage of `super.onCompleted` in the documentation examples. I'll check it and update this part if possible.

And I tried calling `request.cancel` (`clientCallStreamObserver.cancel("Reason Here", null)`) but it didn't work well, because when we cancel through this method, it invokes `onError` from this object.

Reply:
After checking how you've done it on GitpodManager class, the idea got clearer, so started using `request.cancel`. I've updated the code.